### PR TITLE
Gradient activity analysis as a JAX interpreter

### PR DESCRIPTION
**Pull request by dime10**
This PR introduces a new JAX transform (`jax.core.Trace`) that allows us to track the gradient activity of program values. Activity determines whether a value participates in differentiation, and transitively whether an operation participates in differentiation. The transform itself does nothing (identity transform), but its tracers record the activity of values as they pass through JAX primitives. 

The starting point is the `catalyst.tracing.trace_diffargs` transform, which is applied automatically to any function supplied to Catalyst AD methods (`grad`, `jacobian`, `value_and_grad`, etc). Based on the `argnums` parameter from those high-level API functions, it will initialize tracers with the appropriate activity for the function arguments. Primitives are then interpreted using the next trace in the stack, and the activity of outputs is based on the activity of inputs (in an opaque manner, one could imagine defining "activity rules" for primitives for more fine-grained control).

While the transform generally works with higher-order primitives as well (those that trace sub-functions in a separate tracing frame), by automatically propagating itself to the nested frame, it only works if the primitive is a `CallPrimitive`. Special cases like Catalyst control flow primitives still need to be implemented, especially considering Catalyst's use of customized tracing code, which complicates the matter.

The activity analysis can in principle be used in several ways, but Catalyst will use it to skip verification and decomposition of operators based on differentiability criteria if the op does not participate in differentiation.

The following example now passes Catalyst's verification which it didn't before:
```py
# QNode/QFunc support
@qml.qnode(qml.device("lightning.qubit", wires=0))
def circ(x, y, z):

    # propagate activity through other operations
    psi = jnp.sin(y + 1)

    # as long as x and y are not in argnums, the verification passes
    qml.Rot(x, psi, 0.1, wires=0)

    # pytree support
    return {"hi": qml.expval(qml.Z(0))}

@qjit
def main(x: float, y: float, z: float):
    # Catalyst AD primitives automatically inject the activity analysis transform
    return grad(circ, argnums=[2])(x, y, z)
```

Some TODOs:
- compatibility with control flow primitives
- optimize analysis for call primitives by looking at the activity of results
- additional testing

Note that we don't want to use JAX `JVPTracer`s for this because it will try to interpret all the primitives through their JVP rules, which we don't support for our custom (quantum) primitives, and classical AD is also handled by Catalyst further down.

[sc-84808]

### CODE REVIEW

File: frontend/catalyst/tracing/interpreters.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+def trace_diffargs(fun, argnums):
+
+    @functools.wraps(fun)
+    def diffargs_transform_wrapper(*args, **kwargs):
+        args_flat, args_tree = jax.tree_flatten(args)
+        fun_flat, res_tree = flatten_fun(fun, args_tree)
+
+        with jax.core.new_main(ADTrace) as main:
+            trace = ADTrace(main, jax.core.cur_sublevel())
```
Comment (mudit2812) [inline frontend/catalyst/tracing/interpreters.py:101]:
```suggestion
            trace = main.with_cur_sublevel()
```

File: frontend/test/pytest/test_verification.py [hunk 2]
```diff
@@ -910,6 +912,48 @@ def cir(x: float):
...
+            return grad(circ, argnums=argnums)(x, y, z)
+
+        with pytest.raises(DifferentiableCompileError, match="Rot is non-differentiable"):
+            qjit(target="mlir")(main)
+
+    @pytest.mark.parametrize("argnums", (2,))
+    def test_inactive_argument_verification(self, argnums):
+        """Test that verification is skipped correctly if all arguments are inactive."""
+
+        @qml.qnode(qml.device("lightning.qubit", wires=0))
```
Comment (paul0403) [inline frontend/test/pytest/test_verification.py:943]: Is there a reason why `wires=0` here, but `wires=1` above?

File: frontend/catalyst/tracing/interpreters.py (hunk 1 above)
Comment (dime10) [reply frontend/catalyst/tracing/interpreters.py:101]: Are they not equivalent? I kinda like the explicit version :) 

As a side note, I just realized `MainTrace` is actually gone from the latest JAX main branch (and with it this function) 😱

File: frontend/test/pytest/test_verification.py (hunk 2 above)
Comment (dime10) [reply frontend/test/pytest/test_verification.py:943]: Slight mistake
```suggestion
        @qml.qnode(qml.device("lightning.qubit", wires=1))
```

File: frontend/catalyst/tracing/interpreters.py (hunk 1 above)
Comment (mudit2812) [reply frontend/catalyst/tracing/interpreters.py:101]: > Are they not equivalent? I kinda like the explicit version :)

Yep, the same. I like `MainTrace.with_cur_sublevel()` better because if you were to add additional args to your trace, you can specify them as the `payload` for `new_main`, and it automatically populates those arguments when  using `with_cur_sublevel()`. You don't have any args in `ADTrace` that are not present in the base class so it's fine, but if you make any changes to this in the future, it would be one less minor but crucial detail to worry about. That being said, you don't have to worry about this at all if we bump JAX.

> As a side note, I just realized MainTrace is actually gone from the latest JAX main branch (and with it this function) 😱

Yep, now they use a much more isolated stack implementation. Rather than having an actual stack, they have a linked-list style implementation. At any given time, a trace knows its parent (not literally, JAX uses context managers to effectively achieve this), and that's about it.

Comment (dime10) [reply frontend/catalyst/tracing/interpreters.py:101]: Thanks! :)

File: frontend/catalyst/jax_tracer.py [hunk 3]
```diff
@@ -1150,6 +1152,8 @@ def trace_quantum_function(
         out_tree: PyTree shapen of the result
     """
 
+    # This generates a new base main (level 0) upon entering, which is separate from the existing
+    # stack? (TODO: deduplicate?)
```
Comment (paul0403) [inline frontend/catalyst/jax_tracer.py:1156]: Is it possible to not have the new `ADTrace` interpreter as a new main trace, but append it into the list of interpreters we already have? I don't know exactly how it's done, but from @mudit2812 's comment

> Yep, now they use a much more isolated stack implementation. Rather than having an actual stack, they have a linked-list style implementation. At any given time, a trace knows its parent (not literally, JAX uses context managers to effectively achieve this), and that's about it.

it does sound possible?

**Review by paul0403 — COMMENTED** [review]
Thanks @dime10 :muscle: ! My knowledge about tracing is quite limited but I left some reviews. Maybe someone more experienced in these things can chime in as well.

File: frontend/catalyst/tracing/interpreters.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+class ADTrace(jax.core.Trace):
+    def pure(self, x):
+        return ADTracer(self, x, active=False)
+
+    def lift(self, x):
+        return ADTracer(self, x, active=False)
+
+    def process_primitive(self, primitive, in_tracers, params):
+        # Propagate differentiability information. This should be the conservative action.
+        # TODO: Improve the algorithm by taking into account semantics of individual primitives.
```
[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/tracing/interpreters.py:36]: That wouldn't work. The interpreter has to propagate the activity of values through _all_ primitives (or at least those that receive an ADTracer as input), otherwise the information is lost as soon you encounter non-quantum primitives.

File: frontend/catalyst/tracing/interpreters.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+        is_active = any(tracer.active for tracer in in_tracers)
```
[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/tracing/interpreters.py:37]: Yeah you're right, to understand the code fully unfortunately one also needs to know what JAX functions are doing ([Autodidax](https://docs.jax.dev/en/latest/autodidax.html) is a great resource for this, but it really does take time to wrap ones head around it). In short, the inputs to a Trace's `process_primitive` function are always "raised" to that Trace's level (definition of `Primitive.bind` from the autodidax):
```py
def bind(prim, *args, **params):
  top_trace = find_top_trace(args)
  tracers = [full_raise(top_trace, arg) for arg in args]
  outs = top_trace.process_primitive(prim, tracers, params)
  return [full_lower(out) for out in outs]
```

A consequence of their architecture is that a Trace is never invoked if none of the primitives inputs are of the Trace's tracer type, since it won't be produced by the `find_top_trace` function. In order to ensure our Trace is invoked for the next primitive again, we have to box up at least one of the results of the current primitive into our tracer type.
Note there is also the `full_lower` call, which will unbox our tracers, but only if it makes sense to do so (e.g. a `ADTracer(..., active=False)` can safely be converted to a lower-level tracer or constant.

File: frontend/catalyst/tracing/interpreters.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+        argnums = [idx for idx, x in enumerate(in_tracers) if x.active]
+        f_wrapped = lu.WrappedFun(
+            trace_diffargs(f.f, argnums), f.transforms, f.stores, f.params, f.in_type, f.debug_info
+        )
+
+        lowered_tracers = [tracer.value for tracer in in_tracers]
+        out_tracers = call_primitive.bind(f_wrapped, *lowered_tracers, **params)
+
+        # TODO: Extract the differentiability of the call results.
+        assert call_primitive.multiple_results
```
[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/tracing/interpreters.py:67]: It's more of a marker that a primitive _can_ produce multiple (or none) results, and it does not change instance by instance (see also how we use this field in our primitives). You can think of it more like a class attribute, and for `CallPrimitive`s this should [always be true ](https://github.com/jax-ml/jax/blob/f9f47217df9705520c05ae8ea718a6ebb8197fd2/jax/_src/core.py#L2428-L2429) (unless someone introduces some weird call primitive and purposefully sets this attribute to false).

File: frontend/catalyst/tracing/interpreters.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+        return [ADTracer(self, x, bool(argnums)) for x in out_tracers]
```
[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/tracing/interpreters.py:68]: Yes, only problem is we need to gather that information during the nested jaxpr tracing, we can't just look at the results of the call primitive itself.

File: frontend/catalyst/jax_tracer.py (hunk 3 above)
Comment (dime10) [reply frontend/catalyst/jax_tracer.py:1156]: This comment doesn't relate to the `ADTrace` I wrote, as I was debugging our current tracing code I discovered all sorts of weird behaviour, like doubling on Traces in the stack unnecessarily, and this one creating an entirely new _base main_ (note a "base main" is not the same as a "regular main" trace, all JAX transforms are instantiated as new "main traces"). The new base main essentially instantiates a new Trace stack, but then we append to existing one anyway so I don't know my Sergei chose this path. Maybe it's an oversight and it can be removed, not sure 🤔 

The new changes in JAX will be another thing to consider, but right now this Trace is following the way JAX does things in the release we currently depend on (and autodidax).

Comment (mudit2812) [reply frontend/catalyst/jax_tracer.py:1156]: Yep, my comment was strictly in reference to JAX 0.5 and is irrelevant to 0.4.28

File: frontend/catalyst/tracing/interpreters.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+        results = []
+        for tracer in out_tracers:
+            if isinstance(jax.core.get_aval(tracer), (AbstractQreg, AbstractQbit)):
+                results.append(tracer)
+            else:
+                results.append(ADTracer(self, tracer, is_active))
+
+        return results if primitive.multiple_results else results[0]
+
+    def process_call(self, call_primitive, f: lu.WrappedFun, in_tracers, params):
```
Comment (erick-xanadu) [inline frontend/catalyst/tracing/interpreters.py:56]: Hey, these look to be what JAX calls final style interpreters. Would that be correct?

Comment (dime10) [reply frontend/catalyst/tracing/interpreters.py:56]: I always confuse the two 😅 But yes, "final style" is on the fly transformation of primitives 👍

Comment (dime10) [reply frontend/catalyst/tracing/interpreters.py:56]: Fair, but final style is a bit more suited for this transform I think. Compared to the shenanigans we pull in the split-tracing mechanism in the Catalyst frontend, this sort of JAX code is pretty clean imo.

Comment (erick-xanadu) [reply frontend/catalyst/tracing/interpreters.py:56]: That's ok. Just wanting to point this trade-off.

File: frontend/catalyst/tracing/interpreters.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (paul0403) [reply frontend/catalyst/tracing/interpreters.py:36]: Makes sense, thanks

File: frontend/catalyst/tracing/interpreters.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (paul0403) [reply frontend/catalyst/tracing/interpreters.py:37]: I see. I suppose the raise is what the `lift` function (which returns `ADTracer`) is doing?

File: frontend/catalyst/tracing/interpreters.py [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+        out_tracers = primitive.bind(*lowered_tracers, **params)
+        out_tracers = (out_tracers,) if not isinstance(out_tracers, (list, tuple)) else out_tracers
+
+        # Since qubit values (tracers) are cached by Catalyst's tracing machinery, it's possible
+        # to produce leaked tracers from any jaxpr transform launched within trace_quantum_function.
+        # Excluding these values from being boxed up in ADTracers will prevent such leaks.
+        results = []
+        for tracer in out_tracers:
+            if isinstance(jax.core.get_aval(tracer), (AbstractQreg, AbstractQbit)):
+                results.append(tracer)
```
[the remark this answers is not in the document]
Comment (paul0403) [reply frontend/catalyst/tracing/interpreters.py:50]: I think adding the floating point criteria makes a lot of sense!

File: frontend/catalyst/tracing/interpreters.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/tracing/interpreters.py:37]: Yes exactly, there are two ways to raise a value:
- through `pure`: this is reserved for non-tracer values
- through `lift`: this is reserved for tracer values from lower levels
Most of the time the methods do the exact same thing, but there the api offers the option to differentiate between the two.

**Comment by dime10** [comment]
Closing in favour of #1569 due to the incompatibility of the custom tracing system with composable transforms.
